ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatSend.java

```diff
@@ -110,44 +112,41 @@ private void calculateOffsets(boolean enableDataPrefetch) throws IOException, Me
-          short version = headerVersion.getShort();
+          BufferedInputStream bufferedInputStream =
+              new BufferedInputStream(new MessageReadSetIndexInputStream(readSet, i, 0),
+                  BUFFERED_INPUT_STREAM_BUFFER_SIZE);
+          // read and verify header version
+          byte[] headerVersionBytes = new byte[Version_Field_Size_In_Bytes];
+          bufferedInputStream.read(headerVersionBytes, 0, Version_Field_Size_In_Bytes);
+          short version = ByteBuffer.wrap(headerVersionBytes).getShort();
           if (!isValidHeaderVersion(version)) {
             throw new MessageFormatException(
                 "Version not known while reading message - version " + version + ", StoreKey " + readSet.getKeyAt(i),
                 MessageFormatErrorCodes.Unknown_Format_Version);
           }
-          ByteBuffer header = ByteBuffer.allocate(getHeaderSizeForVersion(version));
-          // read the header
-          startTime = SystemTime.getInstance().milliseconds();
-          headerVersion.clear();
-          header.putShort(headerVersion.getShort());
-          readSet.writeTo(i, Channels.newChannel(new ByteBufferOutputStream(header)), Version_Field_Size_In_Bytes,
-              header.capacity() - Version_Field_Size_In_Bytes);
-          logger.trace("Calculate offsets, read header time: {}", SystemTime.getInstance().milliseconds() - startTime);
+          logger.trace("Calculate offsets, read and verify header version time: {}",
+              SystemTime.getInstance().milliseconds() - startTime);
 
-          startTime = SystemTime.getInstance().milliseconds();
-          header.flip();
+          // read and verify header
+          byte[] headerBytes = new byte[getHeaderSizeForVersion(version)];
+          bufferedInputStream.read(headerBytes, Version_Field_Size_In_Bytes,
+              headerBytes.length - Version_Field_Size_In_Bytes);
+
+          ByteBuffer header = ByteBuffer.wrap(headerBytes);
+          header.putShort(version);
+          header.rewind();
           MessageHeader_Format headerFormat = getMessageHeader(version, header);
           headerFormat.verifyHeader();
-          int storeKeyRelativeOffset = header.capacity();
 
-          StoreKey storeKey = storeKeyFactory.getStoreKey(
-              new DataInputStream(new MessageReadSetIndexInputStream(readSet, i, storeKeyRelativeOffset)));
+          // read and verify storeKey
+          StoreKey storeKey = storeKeyFactory.getStoreKey(new DataInputStream(bufferedInputStream));
           if (storeKey.compareTo(readSet.getKeyAt(i)) != 0) {
             throw new MessageFormatException(
                 "Id mismatch between metadata and store - metadataId " + readSet.getKeyAt(i) + " storeId " + storeKey,
                 MessageFormatErrorCodes.Store_Key_Id_MisMatch);
           }
-          logger.trace("Calculate offsets, verify header time: {}",
+          logger.trace("Calculate offsets, read and verify header time: {}",
```

Comment:
1. It doesn't look like you re-initialized the start time so this will be read and verify version time + read and verify header time, right?
2. This time also includes time taken to read the store key - should that be a different log message?

Reply:
Previously it looks like we had
1. Time to read version
2. Time to read header
3. Time to verify header + store key.

Can we have something similar?
1. Time to read version
2. Time to read and verify header
3. Time to read store key